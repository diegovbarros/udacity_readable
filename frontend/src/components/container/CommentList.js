import React,{PureComponent} from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types';


import CommentDetail from '../presentational/CommentDetail';

import {Creators as CommentCreators,getCommentsByPost} from '../../store/features/comment'

const propTypes = {
    postId : PropTypes.string.isRequired,
    onHandleSetComment : PropTypes.func.isRequired,
    loading: PropTypes.bool.isRequired
};



/**
* @description 
* Componente que representa a lista de comentários do post carregado
* @param {String} postId                    Id do post atual
* @param {Boolean} loading                  Se é necessário mostrar o loading
* @param {Function} onDeleteComment        Função apaga um comentário do post
* @param {Function} onHandleSetComment      Função que atualiza ou adiciona um novo comentário ao post
*/
class CommentList extends PureComponent {

    componentDidMount() {
        const { postId,getComments} = this.props;
        getComments(postId)
    }

    /**
    * @description Seta o comment a ser atualizado e enviado ao form com o parentId do post em questão.
    *
    * @param   {Object} comment  Comentário
    * @param   {Event} event     Evento
    */
    onBindComment = (comment,event) => {
        event.preventDefault()
        const { postId,onHandleSetComment} = this.props;
        onHandleSetComment({
            ...comment,parentId : postId
        }); 
    }

    render() {
        const { comments,authUser,onDeleteComment,voteComment,loading} = this.props
        return (
            <div className="ui-block">
                <div className="crumina-module crumina-heading with-title-decoration">
                    <h5 className="heading-title">Comments ({comments.length})</h5>
                </div>
                <ul className="comments-list">
                    {comments && authUser
                        && 
                        comments.map(comment => (
                            <CommentDetail 
                                key={comment.id} 
                                comment={comment} 
                                authUser={authUser} 
                                onDeleteComment={onDeleteComment} 
                                onBindComment={this.onBindComment} 
                                onVoteComment={voteComment}
                                loading={loading}
                            />
                        ))
                    }
                </ul>
            </div>
        )
    }

}

function mapStateToProps (state,ownProps) {
    const {user}    = state;
    const {postId}  = ownProps;
    return {
        comments : getCommentsByPost(postId)(state),
        authUser : user
    }
}
    
function mapDispatchToProps (dispatch) {
    return {
        voteComment: (id,user,option,event) => {
            event.preventDefault();
            dispatch(CommentCreators.vote(id,user,option))
        },
        getComments: (id)              =>  {
            dispatch(CommentCreators.fetch(id));
        },
    }
}


CommentList.propTypes    = propTypes;

export default connect(mapStateToProps,mapDispatchToProps)(CommentList)