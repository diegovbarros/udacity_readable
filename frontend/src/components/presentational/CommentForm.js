import React,{PureComponent} from 'react'
import PropTypes from 'prop-types';
import {genUUID} from '../../utilities/helpers'
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles';
import green from '@material-ui/core/colors/green';
import CircularProgress from '@material-ui/core/CircularProgress';

const styles = {
  buttonNormal: {
        backgroundColor: '#38a9ff',
        '&:hover': {
          backgroundColor: '#38a9ff',
        },
        width: '100%',
        height: '50px'
    },
    buttonProgress: {
        color: green[500],
        position: 'absolute',
        top: '80%',
        left: '47%'
      },
  };

  

const propTypes = {
    comment : PropTypes.shape({
        id: PropTypes.string,
        parentId : PropTypes.string,
        timestamp: PropTypes.number,
        body: PropTypes.string.isRequired,
        author: PropTypes.string,
        voteScore: PropTypes.number
      }),
      user : PropTypes.shape({
        name: PropTypes.string.isRequired
      }).isRequired,
      postId               : PropTypes.string.isRequired,
      onHandleComment      : PropTypes.func.isRequired,
      classes: PropTypes.object.isRequired,
      loading: PropTypes.bool.isRequired
};

const defaultProps  = {
    comment : {
        body : ""
    },
    user : {
        name: ""
    }
};

/**
* @description 
* Componente que representa o form de cadastrou ou atualização de um comment
* @param {String} comment                   Comment a ser atualizado
* @param {String} postId                    Id do post a ser inserido o comentário
* @param {Boolean} loading                  Se é necessário mostrar o loading
* @param {Function} onHandleComment         Método responsável por adicionar/atualizar um comentário
*/
class CommentForm extends PureComponent {

    state = {
        comment : this.props.comment
    }

    componentWillReceiveProps(props) {
         this.setState({comment : props.comment}) ;
    }
    
    /**
    * @description 
    * Faz o bind do objeto comentário para realizar a inserção ou atualização
    * @param {Event} event  Evento do click do botão
    */
    bindComment(event){
        event.preventDefault();
        
        const {onHandleComment,user,postId} = this.props
        const {comment}                     = this.state
        const is_new                        = comment.id ?(false) :(true)
        const id                            = is_new ?(genUUID()) :(comment.id)
        const update                        = Object.assign(comment,
            {
                timestamp : + new Date(),
                author : user.name,
                parentId : postId,
                id,
                is_new
            }
        );
        this.setState({comment : {body: ""}})
        onHandleComment(update)
    }

    /**
    * @description 
    * Recebe o evento que representa a digitação do body do comment
    * 
    * @param   {Event} event Evento 
    */
    onInputSearchChange = (event) => {
        event.preventDefault();
        const {comment} = this.state
        const body =  event.target.value
        console.log(body)
        this.setState({comment : {
            ...comment,
            ...{body:event.target.value}
        }}) ;
    };

    render() {
        
        const {user,classes,loading}            = this.props
        const {comment}                         = this.state
        const enabled                           = comment.body.length <= 0 || loading
       
        return (
            <div className="ui-block">
                <div className="crumina-module crumina-heading with-title-decoration">
                    <h5 className="heading-title">Write a Comment</h5>
                </div>
                <div className="row">
                    <div className="col col-12 col-xl-6 col-lg-6 col-md-6 col-sm-12">
                        <div className="form-group label-floating">
                            <label className="control-label">Name</label>
                            <input className="form-control" placeholder="Your Name" readOnly defaultValue={user.name} type="text" />
                        </div>
                    </div>
                    <div className="col col-12 col-xl-12 col-lg-12 col-md-12 col-sm-12">
                        <div className="form-group label-floating">
                            <label className="control-label">Comment</label>
                            <textarea className="form-control" placeholder="Your Comment" value={comment.body} onChange={this.onInputSearchChange}>
                               
                            </textarea>
                        </div>
                        <Button
                            variant="contained"
                            color="primary"
                            className={classes.buttonNormal}
                            onClick={(event) => this.bindComment(event)}
                            disabled={enabled}>
                            Post your Comment
                        </Button>
                        {loading && <CircularProgress size={24} className={classes.buttonProgress} />}
                    </div>
                </div>
            
            </div>
        )
    }

}


CommentForm.propTypes    = propTypes;
CommentForm.defaultProps = defaultProps;

export default withStyles(styles)(CommentForm)