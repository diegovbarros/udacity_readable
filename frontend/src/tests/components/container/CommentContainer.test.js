import React from 'react';
import ReactRouterEnzymeContext from 'react-router-enzyme-context';
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import CommentContainerConnect ,{CommentContainer}from '../../../components/container/CommentContainer';
import CommentList from '../../../components/container/CommentList';
import CommentForm from '../../../components/presentational/CommentForm';

import commentsMock from '../../mocks/comments';

describe('[Container Component] CommentList', () => {
  const context 	= new ReactRouterEnzymeContext();
	const comments 	= Object.keys(commentsMock).map((id) => commentsMock[id])
	let props 			= {}

	beforeEach(() => {
		props = {
			postId: "8xf0y6ziyjabvozdd253nd",
			user : {name: "Udacity",avatar: "udacity"},
			app:{loading: false},
			comments
		};
	})

  it('Shallow renders correctly', () => {
    expect(global.shallow(<CommentContainer {...props} />,context.get()));
	});


	it('Shallow renders correctly', () => {

		const mockStore = configureMockStore([thunk]);
		const store = mockStore({
			user: {name:"Udacity",avatar: "udacity"}
		});

		expect(global.shallow(<CommentContainerConnect store={store} postId={"sa"} />,context.get()));
	});

	it('Renders correctly', () => {
    expect(global.shallow(<CommentContainer {...props} />)).toMatchSnapshot();
	});

	it('Shallow renders correctly', () => {

		const mockStore = configureMockStore([thunk]);
		const store = mockStore({
			app: {}
		});

		expect(global.shallow(<CommentContainerConnect store={store} {...props} />)).toMatchSnapshot();
	});

	it('Expect to find CommentList and CommentForm nested', () => {
		const wrapper = global.shallow(<CommentContainer {...props} />);
		expect(wrapper.find(CommentList)).toHaveLength(1);
		expect(wrapper.find(CommentForm)).toHaveLength(1);
	});




});
