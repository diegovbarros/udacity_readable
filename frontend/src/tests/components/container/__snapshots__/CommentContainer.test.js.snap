// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Container Component] CommentList Renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CommentContainer
    app={
        Object {
            "loading": false,
          }
    }
    comments={
        Array [
            Object {
              "author": "Udacity",
              "body": "Hi there! I am a COMMENT.",
              "deleted": false,
              "id": "894tuq4ut84ut8v4t8wun89g",
              "parentDeleted": false,
              "parentId": "8xf0y6ziyjabvozdd253n",
              "timestamp": 1468166872634,
              "votes": Array [],
            },
            Object {
              "author": "Udacity",
              "body": "Comments. Are. Cool.",
              "deleted": false,
              "id": "8tu4bsun805n8un48ve89",
              "parentDeleted": false,
              "parentId": "8xf0y6ziyjabvozdd253nd",
              "timestamp": 1469479767190,
              "votes": Array [],
            },
          ]
    }
    postId="8xf0y6ziyjabvozdd253nd"
    user={
        Object {
            "avatar": "udacity",
            "name": "Udacity",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Unknown>
        <Connect(CommentList)
                loading={false}
                onDeleteComment={[Function]}
                onHandleSetComment={[Function]}
                postId="8xf0y6ziyjabvozdd253nd"
        />
        <WithStyles(CommentForm)
                comment={undefined}
                loading={false}
                onHandleComment={[Function]}
                postId="8xf0y6ziyjabvozdd253nd"
                user={
                        Object {
                                "avatar": "udacity",
                                "name": "Udacity",
                              }
                }
        />
</Unknown>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(CommentList)
            loading={false}
            onDeleteComment={[Function]}
            onHandleSetComment={[Function]}
            postId="8xf0y6ziyjabvozdd253nd"
/>,
          <WithStyles(CommentForm)
            comment={undefined}
            loading={false}
            onHandleComment={[Function]}
            postId="8xf0y6ziyjabvozdd253nd"
            user={
                        Object {
                                    "avatar": "udacity",
                                    "name": "Udacity",
                                  }
            }
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "loading": false,
            "onDeleteComment": [Function],
            "onHandleSetComment": [Function],
            "postId": "8xf0y6ziyjabvozdd253nd",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "comment": undefined,
            "loading": false,
            "onHandleComment": [Function],
            "postId": "8xf0y6ziyjabvozdd253nd",
            "user": Object {
              "avatar": "udacity",
              "name": "Udacity",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Unknown>
          <Connect(CommentList)
                    loading={false}
                    onDeleteComment={[Function]}
                    onHandleSetComment={[Function]}
                    postId="8xf0y6ziyjabvozdd253nd"
          />
          <WithStyles(CommentForm)
                    comment={undefined}
                    loading={false}
                    onHandleComment={[Function]}
                    postId="8xf0y6ziyjabvozdd253nd"
                    user={
                              Object {
                                        "avatar": "udacity",
                                        "name": "Udacity",
                                      }
                    }
          />
</Unknown>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Connect(CommentList)
              loading={false}
              onDeleteComment={[Function]}
              onHandleSetComment={[Function]}
              postId="8xf0y6ziyjabvozdd253nd"
/>,
            <WithStyles(CommentForm)
              comment={undefined}
              loading={false}
              onHandleComment={[Function]}
              postId="8xf0y6ziyjabvozdd253nd"
              user={
                            Object {
                                          "avatar": "udacity",
                                          "name": "Udacity",
                                        }
              }
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "loading": false,
              "onDeleteComment": [Function],
              "onHandleSetComment": [Function],
              "postId": "8xf0y6ziyjabvozdd253nd",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "comment": undefined,
              "loading": false,
              "onHandleComment": [Function],
              "postId": "8xf0y6ziyjabvozdd253nd",
              "user": Object {
                "avatar": "udacity",
                "name": "Udacity",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": Symbol(react.fragment),
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`[Container Component] CommentList Shallow renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withRouter(Connect(CommentContainer))
    app={
        Object {
            "loading": false,
          }
    }
    comments={
        Array [
            Object {
              "author": "Udacity",
              "body": "Hi there! I am a COMMENT.",
              "deleted": false,
              "id": "894tuq4ut84ut8v4t8wun89g",
              "parentDeleted": false,
              "parentId": "8xf0y6ziyjabvozdd253n",
              "timestamp": 1468166872634,
              "votes": Array [],
            },
            Object {
              "author": "Udacity",
              "body": "Comments. Are. Cool.",
              "deleted": false,
              "id": "8tu4bsun805n8un48ve89",
              "parentDeleted": false,
              "parentId": "8xf0y6ziyjabvozdd253nd",
              "timestamp": 1469479767190,
              "votes": Array [],
            },
          ]
    }
    postId="8xf0y6ziyjabvozdd253nd"
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    user={
        Object {
            "avatar": "udacity",
            "name": "Udacity",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
